tests/test/functions.rs

```diff
@@ -23,3 +23,42 @@ fn functions_are_sized() {
         }
     }
 }
+
+#[test]
+fn fn_defs() {
+    test! {
+        program {
+            trait Foo { }
+
+            struct Bar { }
+
+            struct Xyzzy { }
+            impl Foo for Xyzzy { }
+
+            fn baz<T>(quux: T) -> T
+                where T: Foo;
+
+            fn garply(thud: i32) -> i32;
+        }
+
+        goal {
+            WellFormed(baz<Bar>)
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            WellFormed(baz<Xyzzy>)
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            WellFormed(garply)
+        } yields {
+            "Unique"
+        }
```

Comment:
Let's add an implied bounds test. Consider:

```
goal {
    FromEnv(baz<Bar>) => Implemented(Bar: Foo)
}
```

which I think should be provable